lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1629,6 +1639,146 @@ struct ConvertLayoutOpConversion
+    auto elemPtrTy =
+        LLVM::LLVMPointerType::get(getTypeConverter()->convertType(elemTy), 3);
+    smemBase = rewriter.create<LLVM::BitcastOp>(loc, elemPtrTy, smemBase);
+    unsigned numWordsEachRep = product<unsigned>(wordsInEachRep);
+    SmallVector<Value> wordVecs(numWordsEachRep);
+    for (unsigned i = 0; i < numElems; ++i) {
+      if (i % srcAccumSizeInThreads == 0) {
+        // start of a replication
+        for (unsigned w = 0; w < numWordsEachRep; ++w) {
+          wordVecs[w] = rewriter.create<LLVM::UndefOp>(loc, wordTy);
+        }
+      }
+      unsigned linearIdxInNanoTile = i % srcAccumSizeInThreads;
+      auto multiDimIdxInNanoTile = getMultiDimIndex<unsigned>(
+          i % srcAccumSizeInThreads, srcBlockedLayout.getSizePerThread());
+      multiDimIdxInNanoTile[inOrd[0]] /= minVec;
+      unsigned pos = multiDimIdxInNanoTile[inOrd[0]] % minVec;
+      unsigned wordVecIdx =
+          getLinearIndex<unsigned>(multiDimIdxInNanoTile, wordsInEachRep);
+      wordVecs[wordVecIdx] = rewriter.create<LLVM::InsertElementOp>(
+          loc, wordTy, wordVecs[wordVecIdx], inVals[i],
+          createIndexConst(rewriter, loc, pos));
+
+      if (i % srcAccumSizeInThreads == srcAccumSizeInThreads - 1) {
+        // end of replication, store the vectors into shared memory
+        unsigned linearRepIdx = i / srcAccumSizeInThreads;
+        auto multiDimRepIdx = getMultiDimIndex<unsigned>(linearRepIdx, reps);
+        for (unsigned linearWordIdx = 0; linearWordIdx < numWordsEachRep;
+             ++linearWordIdx) {
+          // step 1: recover the multidim_index from the index of input_elements
+          // row = multiDimOffsetFirstElem[0] +
+          //       multiDimRepIdx[0] * shapePerCTA[0] + multiDimWordIdx[0] *
+          //       (inOrd[0] == 0) ? minVec : 1
+          // col = multiDimOffsetFirstElem[1] +
+          //       multiDimRepIdx[1] * shapePerCTA[1] + multiDimWordIdx[1] *
+          //       (inOrd[0] == 1) ? minVec : 1
+          auto multiDimWordIdx =
+              getMultiDimIndex<unsigned>(linearWordIdx, wordsInEachRep);
+          SmallVector<Value> multiDimIdx(2);
+          multiDimIdx[0] = rewriter.create<LLVM::AddOp>(
+              loc, multiDimOffsetFirstElem[0],
+              createIndexConst(rewriter, loc,
+                               multiDimRepIdx[0] * srcShapePerCTA[0] +
+                                   multiDimWordIdx[0] *
+                                       (inOrd[0] == 0 ? minVec : 1)));
+          multiDimIdx[1] = rewriter.create<LLVM::AddOp>(
+              loc, multiDimOffsetFirstElem[1],
+              createIndexConst(rewriter, loc,
+                               multiDimRepIdx[1] * srcShapePerCTA[1] +
+                                   multiDimWordIdx[1] *
+                                       (inOrd[0] == 1 ? minVec : 1)));
+
+          // step 2: do swizzling
+          Value remained = rewriter.create<LLVM::URemOp>(
```

Comment:
ditto macros for all the rewriter ops in this block; This should go a long way to make the code much more readable